lib/PerfMeter.js

```diff
@@ -0,0 +1,143 @@
+/* globals Services */
+const {Cu} = require("chrome");
+Cu.import("resource://gre/modules/Services.jsm");
+
+const tabs = require("sdk/tabs");
+const simplePrefs = require("sdk/simple-prefs");
+
+const VALID_TELEMETRY_TAGS = new Set([
+  "TAB_READY",
+  "WORKER_ATTACHED",
+  "TOP_FRECENT_SITES_REQUEST",
+  "TOP_FRECENT_SITES_RESPONSE",
+  "RECENT_BOOKMARKS_REQUEST",
+  "RECENT_BOOKMARKS_RESPONSE",
+  "RECENT_LINKS_REQUEST",
+  "RECENT_LINKS_RESPONSE",
+  "FRECENT_LINKS_REQUEST",
+  "FRECENT_LINKS_RESPONSE",
+  "SEARCH_STATE_REQUEST",
+  "SEARCH_STATE_RESPONSE",
+  "NOTIFY_PERFORMANCE",
+]);
+
+function PerfMeter(trackableURLs) {
+  this._trackableURLs = trackableURLs;
+  this.init();
+}
+
+PerfMeter.prototype = {
+
+  _tabs: {},
+
+  get events() {
+    return this._tabs;
+  },
+
+  init: function() {
+    this.onOpen = this.onOpen.bind(this);
+    this.onReady = this.onReady.bind(this);
+    this.onClose = this.onClose.bind(this);
+    this.onPrefChange = this.onPrefChange.bind(this);
+    if (simplePrefs.prefs.performanceLog) {
+      tabs.on("open", this.onOpen);
+      this._active = true;
+    }
+    simplePrefs.on("performanceLog", this.onPrefChange);
+  },
+
+  uninit: function() {
+    if (simplePrefs.prefs.performanceLog) {
+      tabs.removeListener("open", this.onOpen);
+    }
+    this.clearTabs();
+    simplePrefs.removeListener("performanceLog", this.onPrefChange);
+  },
+
+  clearTabs: function() {
+    // remove tab listeners
+    Object.keys(this._tabs).forEach(id => {
+      this._tabs[id].tab.removeListener("ready", this.onReady);
+      this._tabs[id].tab.removeListener("close", this.onClose);
+    });
+    this._tabs = {};
+  },
+
+  onPrefChange: function() {
+    if (simplePrefs.prefs.performanceLog) {
+      tabs.on("open", this.onOpen);
+      this._active = true;
+    } else {
+      tabs.removeListener("open", this.onOpen);
+      this._active = false;
+      this.clearTabs();
+    }
+    Services.obs.notifyObservers(null, "performance-pref-changed", null);
+  },
+
+  isActivityStreamsURL: function(URL) {
+    return this._trackableURLs.indexOf(URL) !== -1;
+  },
+
+  onReady: function(tab) {
+    if (this.isActivityStreamsURL(tab.url)) {
+      this.log(tab, "TAB_READY");
+    } else {
+      // not an activity stream tab, get rid of it
+      delete this._tabs[tab.id];
+    }
+    tab.removeListener("ready", this.onReady);
```

Comment:
Yes, it does.  Note that the tab.removeListener(...) removes the listener attached to the tab passed in.  The "ready" listeners are attached to individual tabs not the tabs object.